File: go/146. LRU Cache.go
type LRUCache struct {
	capacity int
	pointers map[int]*Node
	head, tail *Node
}

type Node struct {
	key, val int
	next, prev *Node
}


func Constructor(capacity int) LRUCache {
	return LRUCache {
		capacity,
		make(map[int]*Node),
		nil,
		nil,
	}
}


func (this *LRUCache) Get(key int) int {
	if _, ok := this.pointers[key]; !ok {
		return -1
	}
	node := this.pointers[key]
	this.makeHead(node)
	return (*node).val
}


func (this *LRUCache) Put(key int, value int)  {
	if _, ok := this.pointers[key]; !ok {
		this.pointers[key] = &Node{key, value, nil, nil}
	}
	node := this.pointers[key]
	node.val = value
	if this.tail == nil {
		this.tail = node
	}
	this.makeHead(node)
	if len(this.pointers) > this.capacity {
		delete(this.pointers, this.removeTail().key)
	}
}

func (this *LRUCache) makeHead(node *Node) {
	if this.head == nil || node == this.head {
		this.head = node
		return
	}
	this.freeNode(node)
	node.next = this.head
	this.head.prev = node
	this.head = node
}

func (this *LRUCache) freeNode(node *Node) {
	if this.tail == node {
		this.tail = this.tail.prev
	}
	if node.prev != nil {
		node.prev.next = node.next
	}
	if node.next != nil {
		node.next.prev = node.prev
	}
	node.next = nil
	node.prev = nil
}

func (this *LRUCache) removeTail () *Node {
	tailNode := this.tail
	this.tail = tailNode.prev
	this.tail.next = nil
	return tailNode
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: java/146. LRU Cache.java
class LRUCache {
    private Node head;
    private Node tail;
    private int capacity;
    private Map<Integer, Node> pointers;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.pointers = new HashMap<>();
        this.head = null;
        this.tail = null;
    }
    
    public int get(int key) {
        if (!pointers.containsKey(key)) return -1;
        Node node = pointers.get(key);
        makeHead(node);
        return node.val;
    }
    
    public void put(int key, int value) {
        Node node = pointers.computeIfAbsent(key, k -> new Node(key, value));
        node.val = value;
        if (tail == null) tail = node;
        makeHead(node);
        if (pointers.size() > capacity) {
            pointers.remove(removeTail().key);
        }
    }

    private void makeHead(Node node) {
        if (head == node || head == null) {
            head = node;
            return;
        }
        makeFree(node);
        node.next = head;
        head.prev = node;
        head = node;
    }

    private Node removeTail() {
        Node tailNode = tail;
        tail = tailNode.prev;
        tail.next = null;
        return tailNode;
    }

    private void makeFree(Node node) {
        if (tail == node) {
            tail = tail.prev;
        }
        if (node.prev != null) {
            node.prev.next = node.next;
        }
        if (node.next != null) {
            node.next.prev = node.prev;
        }
        node.next = null;
        node.prev = null;
    }
}

class Node {
    int key;
    int val;
    Node next;
    Node prev;
    Node (int key, int val) {
        this.key = key;
        this.val = val;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */